Adding IAM Action name trait to override using the API operation name

0xjjoyy opened the pull request:
*Issue #, if available:*

*Description of changes:*
Adds the ActionName trait to override using the API operation name as the IAM action name. Also, allows adding additional actions using the input member field if the value is not null in the request.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

kstich commented on smithy-aws-iam-traits/.gitignore:
```diff
@@ -0,0 +1,3 @@
+gradle/
```
Can remove this, the top level `.gitignore` should handle these.

0xjjoyy replied:
done

kstich commented on smithy-aws-iam-traits/src/main/resources/META-INF/services/software.amazon.smithy.model.traits.TraitService:
```diff
@@ -5,3 +5,4 @@ software.amazon.smithy.aws.iam.traits.DisableConditionKeyInferenceTrait$Provider
 software.amazon.smithy.aws.iam.traits.RequiredActionsTrait$Provider
 software.amazon.smithy.aws.iam.traits.SupportedPrincipalTypesTrait$Provider
 software.amazon.smithy.aws.iam.traits.IamResourceTrait$Provider
+software.amazon.smithy.aws.iam.traits.ActionNameTrait$Provider
```
nit: should have a trailing newline.

0xjjoyy replied:
added

kstich commented on smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/ActionNameTrait.java:
```diff
@@ -0,0 +1,35 @@
+    private ActionNameTrait(String action, FromSourceLocation sourceLocation) {
+        super(ID, action, sourceLocation);
+    }
+
```
Can this have an additional simple constructor that takes only the `action` and passes `SourceLocation.NONE`?

0xjjoyy replied:
added

```
    private ActionNameTrait(String action) {
        super(ID, action, SourceLocation.NONE);
    }
```

kstich commented on smithy-aws-iam-traits/src/main/resources/META-INF/smithy/aws.iam.json:
```diff
@@ -264,6 +264,18 @@
+            },
+            "traits": {
+                "smithy.api#trait": {
+                    "selector": ":test(operation, member)"
```
Sounds like this should be `"operation"`, given the documentation. Tests should be updated to reflect this as well.

0xjjoyy replied:
done

kstich commented on smithy-aws-iam-traits/src/test/resources/software/amazon/smithy/aws/iam/traits/actionname-override.smithy:
```diff
@@ -0,0 +1,17 @@
+$version: "1.0"
```
Should be a v2 model.

0xjjoyy replied:
updated

kstich reviewed (changes requested):
* This will need updates to the specification, that'd be in `aws-iam.rst` in the `docs` folder.
* Java files should use the short-form license header (see #1614)

0xjjoyy commented:
Thanks, made updates.

> This will need updates to the specification, that'd be in aws-iam.rst in the docs folder.

updated

> Java files should use the short-form license header (see https://github.com/awslabs/smithy/pull/1614)

updated

kstich commented on docs/source-2.0/aws/aws-iam.rst:
```diff
@@ -392,6 +392,45 @@ deviates from the :ref:`shape name of the shape ID <shape-id>` of the resource.
+Value type
+    ``string``
+
+Operations not annotated with the ``actionName`` trait use the operation name as the action name.
```
Defaults to the :ref:\`shape name of the shape ID <shape-id>` of the targeted ``operation``.

0xjjoyy replied:
updated

kstich commented on docs/source-2.0/aws/aws-iam.rst:
```diff
@@ -392,6 +392,45 @@ deviates from the :ref:`shape name of the shape ID <shape-id>` of the resource.
+
+The following example defines two operations:
+
+* OperationA is not annoated with the ``actionName`` trait, so has the action name ``OperationA`` which is the same as the operation name.
```
* The first `OperationA` should be in double backticks.
* s/annoated/annotated
* "so has the action name ``OperationA`` which is the same as the operation name." -> and resolves the action name of ``OperationA``.

0xjjoyy replied:
updated

kstich commented on docs/source-2.0/aws/aws-iam.rst:
```diff
@@ -392,6 +392,45 @@ deviates from the :ref:`shape name of the shape ID <shape-id>` of the resource.
+The following example defines two operations:
+
+* OperationA is not annoated with the ``actionName`` trait, so has the action name ``OperationA`` which is the same as the operation name.
+* OperationB has the ``actionName`` trait, so has the action name ``OverridingActionName``.
```
`OperationB` should be in double backticks.

0xjjoyy replied:
updated

0xjjoyy commented on docs/source-2.0/aws/aws-iam.rst:
```diff
@@ -392,6 +392,45 @@ deviates from the :ref:`shape name of the shape ID <shape-id>` of the resource.
+.. smithy-trait:: aws.iam#actionName
+.. _aws.iam#actionName-trait:
+
+-------------------------------
```
updated

0xjjoyy commented on docs/source-2.0/aws/aws-iam.rst:
```diff
@@ -392,6 +392,45 @@ deviates from the :ref:`shape name of the shape ID <shape-id>` of the resource.
+-------------------------------
+
+Summary
+    Provides the ability to override the default action name for the operation. By convention, the action name is the same as the operation name.
```
updated

0xjjoyy commented on smithy-aws-iam-traits/src/main/resources/META-INF/smithy/aws.iam.json:
```diff
@@ -264,6 +264,18 @@
+                "smithy.api#trait": {
+                    "selector": "operation"
+                },
+                "smithy.api#documentation": "Provides the ability to override the default action name for the operation. By convention, the action name is the same as the operation name."
```
updated
